boot/bootstate20.go

```diff
@@ -21,330 +21,526 @@ package boot
+	m, err := ebl.GetBootVars("kernel_status")
+	if err != nil {
+		return err
 	}
 
-	sn, err := snap.ParsePlaceInfoFromSnapFileName(modeenv.Base)
+	ks20.kernelStatus = m["kernel_status"]
+
+	return nil
+}
+
+func (ks20 *bootState20Kernel) status() (string, error) {
+	err := ks20.loadBootloader()
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot get boot base revision: modeenv base variable is invalid: %v", err)
+		return "", err
 	}
 
-	return sn, nil, false, nil
+	return ks20.kernelStatus, nil
 }
 
-// TODO:UC20: implement this
-func (s20 *bootState20Base) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
-	return update, nil
+func (ks20 *bootState20Kernel) setTrySnap(sn snap.PlaceInfo) {
+	ks20.tryKernelSnap = sn
 }
 
-// TODO:UC20: implement this with modeenv
-func (s20 *bootState20Base) setNext(nextKernel snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
-	return false, nil, nil
+func (ks20 *bootState20Kernel) setStatus(status string) {
+	ks20.kernelStatus = status
 }
 
-type bootState20Kernel struct{}
+func (ks20 *bootState20Kernel) revisions() (snap.PlaceInfo, snap.PlaceInfo, bool, error) {
+	var bootSn, tryBootSn snap.PlaceInfo
+	var trying bool
+	err := ks20.loadBootloader()
+	if err != nil {
+		return nil, nil, false, err
+	}
 
-func newBootState20(typ snap.Type) bootState {
-	switch typ {
-	case snap.TypeKernel:
-		return &bootState20Kernel{}
-	case snap.TypeBase:
-		return &bootState20Base{}
-	default:
-		panic(fmt.Sprintf("cannot make a bootState20 for snap type %q", typ))
+	// get the kernel for this bootloader
+	bootSn, err = ks20.ebl.Kernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
-}
 
-func (s20 *bootState20Kernel) revisions() (theSnap, theTrySnap snap.PlaceInfo, trying bool, err error) {
-	var bootSnap, tryBootSnap snap.PlaceInfo
+	tryKernel, tryKernelExists, err := ks20.ebl.TryKernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
+	}
 
-	// the trying snap is the one given by the try-kernel.efi symlink, while
-	// the known-good snap is the one given by the kernel.efi symlink
-	// we can determine if we are trying the try snap by the presence of
-	// 1. a try-kernel.efi symlink
-	// 2. kernel_status == trying
+	if tryKernelExists {
+		tryBootSn = tryKernel
+	}
 
-	bl, err := bootloader.Find("", nil)
+	m, err := ks20.ebl.GetBootVars("kernel_status")
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot find any bootloader: %v", err)
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
-	if !ok {
-		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
+	trying = (m["kernel_status"] == "trying")
```

Comment:
we should probably error if the kernel_status is not empty and tryBootSn is nil

Reply:
or maybe not error but make sure in such cases status get cleared by mark successful?

Reply:
I implemented it such that if kernel_status is not empty, but there is not a try snap, it will be cleaned up in commit() after markSuccessful runs.